src/rules/unused_configuration_options.erl

```diff
@@ -0,0 +1,133 @@
+            filename:extension(File) == ".config" orelse filename:extension(File) == ".src",
+            not is_ignored(File)],
+
+    % get just the options (keys) to search usages
+    ConfigOptions = extract_options(ConfigOptionsByFile),
+
+    % get all the options used by the .erl/.hrl files
+    Uses =
+        [UsedOption
+         || {File, AST} <- FilesAndASTs,
+            filename:extension(File) == ".erl" orelse filename:extension(File) == ".hrl",
+            UsedOption <- options_usage(AST, ConfigOptions)],
+
+    % calculate the unused options
+    UnusedOptions = ConfigOptions -- lists:usort(Uses),
+
+    % build resuts
+    [result(File, Option)
+     || {File, Options} <- ConfigOptionsByFile,
+        Option <- Options,
+        lists:member(Option, UnusedOptions)].
+
+%% It receives a file path and returns a list of options
+%% It's prepared for .config and .app.src files, which contain Erlang Terms
+%% If the file cannot be parsed, it will be ignored (like other user's .config files)
+-spec config_options(file:filename(), hank_context:t()) -> [atom()].
+config_options(File, Context) ->
+    case file:consult(File) of
+        {ok, [ErlangTerms]} ->
+            case is_app_src_file(File) of
+                true ->
+                    {application, _AppName, Options} = ErlangTerms,
+                    EnvOptions = proplists:get_value(env, Options, []),
+                    proplists:get_keys(EnvOptions);
+                false ->
+                    config_keys(ErlangTerms, Context)
+            end;
+        _ ->
+            %% error parsing: ignore the file!
+            []
+    end.
+
+config_keys(ConfigTuples, Context) ->
+    [Key
+     || {AppName, Proplist} <- ConfigTuples,
+        lists:member(AppName, hank_context:project_apps(Context)),
+        Key <- proplists:get_keys(Proplist)].
+
+is_app_src_file(File) ->
+    filename:extension(File) == ".src".
+
+extract_options(OptionsByFile) ->
+    lists:usort([Option || {_File, FileOptions} <- OptionsByFile, Option <- FileOptions]).
+
+options_usage(_AST, []) ->
+    [];
+options_usage(AST, Options) ->
+    [Option
+     || Node <- AST,
+        maybe_contain_options(Node),
+        Option <- Options,
+        is_option_used(erl_syntax:type(Node), Option, Node)].
+
+maybe_contain_options(Node) ->
+    case erl_syntax:type(Node) of
+        function ->
+            true;
+        attribute ->
+            case hank_utils:attr_name(Node) of
+                define ->
+                    true;
+                _ ->
+                    false
+            end;
+        _ ->
+            false
+    end.
+
+is_option_used(function, Option, Function) ->
+    hank_utils:function_has_atom(Function, Option);
+is_option_used(attribute, Option, Macro) ->
+    macro_has_atom(Macro, Option).
+
+macro_has_atom(Macro, Option) ->
+    [_, MacroApplication | _] = erl_syntax:attribute_arguments(Macro),
+    case erl_syntax:type(MacroApplication) of
```

Comment:
I have a feeling that this will be not enough. Can we add a test with the following code…

```erlang
-define(A_COMPLEX_MACRO, fun() -> All = application:get_all_env(test_app), proplists:get_value(this_used_key, All) end).
```

…and verify that `this_used_key` is not marked as unused?

It might be worth adding also tests like these…
```erlang
-define(A_COMPLEX_MACRO(App), fun() -> All = application:get_all_env(App), proplists:get_value(another_used_key, All) end).
-define(A_COMPLEX_MACRO(), fun() -> Value = application:get_env(yet_another_key), list_to_binary(Value) end).
```

Reply:
Covered with https://github.com/AdRoll/rebar3_hank/pull/85/commits/764c0c538ca5488767d36d67b8982b0d046327c0